File: typescript-javascript/day05-strings-json-dates/solutions.js
// Day 5 reference solutions.

function normalizeWhitespace(text) {
    return text.trim().split(/\s+/).join(" ");
}

function formatReport(name, score) {
    return `${name} scored ${score.toFixed(1)}%`;
}

function roundTripThroughJSON(data) {
    return JSON.parse(JSON.stringify(data));
}

function getYearFromDate(dateObject) {
    return dateObject.getFullYear();
}

function daysBetween(date1, date2) {
    const millisecondsBetween = date2.getTime() - date1.getTime();
    return millisecondsBetween / (1000 * 60 * 60 * 24);
}

module.exports = { normalizeWhitespace, formatReport, roundTripThroughJSON, getYearFromDate, daysBetween };

File: python/day05-strings-files/solutions.py
"""Day 5 reference solutions."""


def normalize_whitespace(text):
    return " ".join(text.split())


def parse_csv_line(line):
    parts = [p.strip() for p in line.split(",")]
    return [parts[0], float(parts[1]), parts[2]]


def format_report(name, score):
    return f"{name:<10}: {score:>5.1f}%"


def write_lines_to_file(path, lines):
    with open(path, "w", encoding="utf-8") as f:
        for line in lines:
            f.write(line + "\n")


def read_nonempty_lines(path):
    result = []
    with open(path, encoding="utf-8") as f:
        for line in f:
            stripped = line.strip()
            if stripped:
                result.append(stripped)
    return result
